Matchismo/CardMatchingGameViewController.m

```diff
@@ -0,0 +1,292 @@
+		return;
+	}
+
+	NSUInteger chosenCardViewIndex = [self.cards indexOfObject:gestureRecognizer.view];
+	[self.game chooseCardAtIndex:chosenCardViewIndex];
+
+	[UIView transitionWithView:gestureRecognizer.view
+					  duration:0.5
+					   options:UIViewAnimationOptionTransitionFlipFromLeft
+					animations:nil
+					completion:nil];
+
+	[self updateUI];
+
+	[self addStateToHistory];
+
+	if ([self.game isEnded]) {
+		[self saveGame];
+	}
+}
+
+- (void)saveGame
+{
+	NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+
+	NSMutableDictionary *highScores = [[userDefaults dictionaryForKey:(NSString *)USER_DEFAULTS_HIGH_SCORES] mutableCopy];
+	if (!highScores) {
+		highScores = [[NSMutableDictionary alloc] initWithCapacity:2];
+	}
+
+	NSMutableDictionary *gameHighScores = [[highScores objectForKey:[[self class] gameName]] mutableCopy];
+	if(!gameHighScores) {
+		gameHighScores = [[NSMutableDictionary alloc] initWithCapacity:2];
+		gameHighScores[HIGH_SCORES_CATEGORY_SCORE] = [[NSMutableArray alloc] initWithCapacity:1];
+		gameHighScores[HIGH_SCORES_CATEGORY_TIME] = [[NSMutableArray alloc] initWithCapacity:1];
+		highScores[[[self class] gameName]] = gameHighScores;
+	}
+
+	NSMutableArray *scoreCategoryHighScores = [gameHighScores[HIGH_SCORES_CATEGORY_SCORE] mutableCopy];
+	[scoreCategoryHighScores addObject:[NSNumber numberWithInteger:self.game.score]];
+	NSSortDescriptor* sortOrder = [NSSortDescriptor sortDescriptorWithKey: @"self" ascending: NO];
+	[scoreCategoryHighScores sortUsingDescriptors:[NSArray arrayWithObject:sortOrder]];
+	if ([scoreCategoryHighScores count] > 10) {
+		[scoreCategoryHighScores removeLastObject];
+	}
+	gameHighScores[HIGH_SCORES_CATEGORY_SCORE] = scoreCategoryHighScores;
+
+	NSMutableArray *timeCategoryHighScores = [gameHighScores[HIGH_SCORES_CATEGORY_TIME] mutableCopy];
+	[timeCategoryHighScores addObject:[NSNumber numberWithDouble:self.game.gameLasts]];
+	sortOrder = [NSSortDescriptor sortDescriptorWithKey: @"self" ascending: YES];
+	[timeCategoryHighScores sortUsingDescriptors:[NSArray arrayWithObject:sortOrder]];
+	if ([timeCategoryHighScores count] > 10) {
+		[timeCategoryHighScores removeLastObject];
+	}
+	gameHighScores[HIGH_SCORES_CATEGORY_TIME] = timeCategoryHighScores;
+
+	highScores[[[self class] gameName]] = gameHighScores;
+
+	[userDefaults setObject:highScores forKey:(NSString *)USER_DEFAULTS_HIGH_SCORES];
+	[userDefaults synchronize];
+}
+
+- (void)updateUI
+{
+	for (CardView *cardView in self.cards) {
+		NSUInteger cardViewIndex = [self.cards indexOfObject:cardView];
+		Card *card = [self.game cardAtIndex:cardViewIndex];
+		cardView.faceUp = card.isChosen;
+		cardView.enabled = !card.isMatched;
+	}
+
+	self.scoreLabel.text = [NSString stringWithFormat:@"Score: %ld", (long)self.game.score];
+
+	NSMutableAttributedString *lastCards = [[NSMutableAttributedString alloc] initWithString:@""];
+	for (Card* lastCard in self.game.lastCards) {
```

Comment:
Although it is not a requirement and you can use whichever you wish, you could use [enumerateObjectsUsingBlock:](https://developer.apple.com/reference/foundation/nsarray/1415846-enumerateobjectsusingblock) instead of fast enumeration. It result in speedup if you have a large array and it gives you an additional index variable you can also operate with.